crates/accelerate/src/quantum_circuit/circuit_instruction.rs

```diff
@@ -0,0 +1,256 @@
+use pyo3::basic::CompareOp;
+use pyo3::prelude::*;
+use pyo3::types::{PyList, PyTuple};
+use pyo3::{PyObject, PyResult};
+
+/// A single instruction in a :class:`.QuantumCircuit`, comprised of the :attr:`operation` and
+/// various operands.
+///
+/// .. note::
+///
+///     There is some possible confusion in the names of this class, :class:`~.circuit.Instruction`,
+///     and :class:`~.circuit.Operation`, and this class's attribute :attr:`operation`.  Our
+///     preferred terminology is by analogy to assembly languages, where an "instruction" is made up
+///     of an "operation" and its "operands".
+///
+///     Historically, :class:`~.circuit.Instruction` came first, and originally contained the qubits
+///     it operated on and any parameters, so it was a true "instruction".  Over time,
+///     :class:`.QuantumCircuit` became responsible for tracking qubits and clbits, and the class
+///     became better described as an "operation".  Changing the name of such a core object would be
+///     a very unpleasant API break for users, and so we have stuck with it.
+///
+///     This class was created to provide a formal "instruction" context object in
+///     :class:`.QuantumCircuit.data`, which had long been made of ad-hoc tuples.  With this, and
+///     the advent of the :class:`~.circuit.Operation` interface for adding more complex objects to
+///     circuits, we took the opportunity to correct the historical naming.  For the time being,
+///     this leads to an awkward case where :attr:`.CircuitInstruction.operation` is often an
+///     :class:`~.circuit.Instruction` instance (:class:`~.circuit.Instruction` implements the
+///     :class:`.Operation` interface), but as the :class:`.Operation` interface gains more use,
+///     this confusion will hopefully abate.
+///
+/// .. warning::
+///
+///     This is a lightweight internal class and there is minimal error checking; you must respect
+///     the type hints when using it.  It is the user's responsibility to ensure that direct
+///     mutations of the object do not invalidate the types, nor the restrictions placed on it by
+///     its context.  Typically this will mean, for example, that :attr:`qubits` must be a sequence
+///     of distinct items, with no duplicates.
+#[pyclass(
+    freelist = 20,
+    sequence,
+    get_all,
+    module = "qiskit._accelerate.quantum_circuit"
+)]
+#[derive(Clone, Debug)]
+pub struct CircuitInstruction {
+    /// The logical operation that this instruction represents an execution of.
+    pub operation: PyObject,
+    /// A sequence of the qubits that the operation is applied to.
+    pub qubits: Py<PyTuple>,
+    /// A sequence of the classical bits that this operation reads from or writes to.
+    pub clbits: Py<PyTuple>,
+}
+
+#[pymethods]
+impl CircuitInstruction {
+    #[new]
+    pub fn new(
+        py: Python<'_>,
+        operation: PyObject,
+        qubits: Option<&PyAny>,
+        clbits: Option<&PyAny>,
```

Comment:
The Python-space version of this object took `()` as the optional argument.  Can we do the same here in PyO3?

Reply:
Unfortunately it seems that we cannot. PyO3 doesn't appear to offer a way to default arguments to the empty tuple.